File: src/ee/ads/internal/GuardedRewardedAd.cpp
#include "ee/ads/internal/GuardedRewardedAd.hpp"

#include <mutex>

#include <ee/ads/internal/AsyncHelper.hpp>
#include <ee/core/ObserverHandle.hpp>
#include <ee/core/SpinLock.hpp>
#include <ee/core/Task.hpp>

namespace ee {
namespace ads {
using Self = GuardedRewardedAd;

Self::GuardedRewardedAd(const std::shared_ptr<IRewardedAd>& ad)
    : ad_(ad) {
    loading_ = false;
    loaded_ = false;
    displaying_ = false;

    handle_ = std::make_unique<ObserverHandle>();
    handle_->bind(*ad_).addObserver({
        .onLoaded =
            [this] {
                std::unique_lock<SpinLock> lock(*lock_);
                loaded_ = true;
                lock.unlock();

                // Propagation.
                dispatchEvent([](auto&& observer) {
                    if (observer.onLoaded) {
                        observer.onLoaded();
                    }
                });
            },
        .onClicked =
            [this] {
                // Propagation.
                dispatchEvent([](auto&& observer) {
                    if (observer.onClicked) {
                        observer.onClicked();
                    }
                });
            },
    });

    lock_ = std::make_unique<SpinLock>();
}

Self::~GuardedRewardedAd() = default;

void Self::destroy() {
    ad_->destroy();
    handle_->clear();
}

bool Self::isLoaded() const {
    std::scoped_lock<SpinLock> lock(*lock_);
    return loaded_;
}

Task<bool> Self::load() {
    std::unique_lock<SpinLock> lock(*lock_);
    if (loaded_) {
        lock.unlock();
        co_return true;
    }
    if (displaying_) {
        lock.unlock();
        co_return false;
    }
    if (loading_) {
        // Waiting.
        lock.unlock();
        co_return co_await ad_->load();
    }
    loading_ = true;
    lock.unlock();
    auto result = co_await ad_->load();
    lock.lock();
    loading_ = false;
    lock.unlock();
    co_return result;
}

Task<IRewardedAdResult> Self::show() {
    std::unique_lock<SpinLock> lock(*lock_);
    if (not loaded_) {
        lock.unlock();
        co_return IRewardedAdResult::Failed;
    }
    if (loading_) {
        lock.unlock();
        co_return IRewardedAdResult::Failed;
    }
    if (displaying_) {
        // Waiting.
        lock.unlock();
        co_return co_await ad_->show();
    }
    displaying_ = true;
    lock.unlock();
    auto result = co_await ad_->show();
    lock.lock();
    displaying_ = false;
    if (result == IRewardedAdResult::Failed) {
        // Failed to show, can use this ad again.
    } else {
        loaded_ = false;
    }
    lock.unlock();
    co_return result;
}
} // namespace ads
} // namespace ee

File: unity/Ads/Runtime/EE/Internal/GuardedRewardedAd.cs
using System.Threading.Tasks;

namespace EE.Internal {
    public class GuardedRewardedAd
        : ObserverManager<IRewardedAdObserver>, IRewardedAd {
        private readonly IRewardedAd _ad;
        private bool _loading;
        private bool _displaying;
        private readonly ObserverHandle _handle;

        public GuardedRewardedAd(IRewardedAd ad) {
            _ad = ad;
            _handle = new ObserverHandle();
            _handle.Bind(ad).AddObserver(new IRewardedAdObserver {
                OnLoaded = () => {
                    IsLoaded = true;
                    DispatchEvent(observer => observer.OnLoaded?.Invoke());
                },
                OnClicked = () => DispatchEvent(observer => observer.OnClicked?.Invoke())
            });
        }

        public void Destroy() {
            _ad.Destroy();
            _handle.Clear();
        }

        public bool IsLoaded { get; private set; }

        public async Task<bool> Load() {
            if (IsLoaded) {
                return true;
            }
            if (_displaying) {
                return false;
            }
            if (_loading) {
                // Waiting.
                return await _ad.Load();
            }
            _loading = true;
            var result = await _ad.Load();
            _loading = false;
            return result;
        }

        public async Task<IRewardedAdResult> Show() {
            if (!IsLoaded) {
                return IRewardedAdResult.Failed;
            }
            if (_loading) {
                return IRewardedAdResult.Failed;
            }
            if (_displaying) {
                // Waiting.
                return await _ad.Show();
            }
            _displaying = true;
            var result = await _ad.Show();
            _displaying = false;
            if (result == IRewardedAdResult.Failed) {
                // Failed to show, can use this ad again.
            } else {
                IsLoaded = false;
            }
            return result;
        }
    }
}